pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java

```diff
@@ -202,6 +209,20 @@ private void setupHelixSystemProperties() {
             CommonConstants.Helix.DEFAULT_FLAPPING_TIME_WINDOW_MS));
   }
 
+  private void setupHelixClusterConstraints() {
+    String maxMessageLimit = _config.getString(CommonConstants.Helix.CONFIG_OF_HELIX_INSTANCE_MAX_STATE_TRANSITIONS,
+        CommonConstants.Helix.DEFAULT_HELIX_INSTANCE_MAX_STATE_TRANSITIONS);
+    Map<ClusterConstraints.ConstraintAttribute, String> constraintAttributes = new HashMap<>();
+    constraintAttributes.put(ClusterConstraints.ConstraintAttribute.INSTANCE, ".*");
+    constraintAttributes
+        .put(ClusterConstraints.ConstraintAttribute.MESSAGE_TYPE, Message.MessageType.STATE_TRANSITION.name());
+    ConstraintItem constraintItem = new ConstraintItem(constraintAttributes, maxMessageLimit);
+
+    _helixControllerManager.getClusterManagmentTool()
```

Comment:
While this may be a good thing to do, we have often experienced issues with helix, especially the undocumented features. Can we at least make this so that if the config is not set, then we leave the behavior like before?

Meanwhile. I have asked someone from the helix team to take a look at this.

Reply:
Sure. Please have the helix team take a look.

There is some caveat of not adding the constraint when the config is not set: do we need to add logic removing the constraint if the config is not present? Otherwise it's possible that the constraint is left there, when users remove this configuration. What about if there are inconsistent configurations, with the configuration set on one controller but not others (which will lead to race condition)? Usually having the configuration with a default value makes such issues more manageable.

Reply:
@mcvsubbu with Junkai's assessment, do you still have concerns on the current approach, given it's a safe change for better zookeeper scalability? 
I prefer the current way since it simplifies the configuration story, i.e. no need to use external Helix admin to clean up constraint etc.